### ISSUE TITLE
Extending the `#expect_offense` treatment to auto-correct

### ISSUE THREAD

**Issue by Drenmi** [rubocop-hq/rubocop#4507]
This is aimed at @backus. I don't think I can overstate how much I love the new `#expect_offense` and `#expect_no_offenses` matchers. 😊 Because of how much I love them, however, it always irks me when I need to mix them with old school auto-correct assertions.

I've been thinking about how we could extend the matcher to assert a certain correction--or no correction--in cases where it applies. (I apologize for the stream-of-consciousness style writing below. None of these ideas are very well formed. I mostly want to get the discussion going. 🙂)

---

At first, I was thinking something along the lines of:

```ruby
expect_offense(<<-RUBY.strip_indent, <<-CORRECTION.strip_indent)
  foo
RUBY
  bar
CORRECTION
```

This has it's own problems though. You can't use `RUBY` delimiter for both heredocs, so we forego the highlighting benefits. In addition, explicitly asserting "no correction" requires duplicating the contents of the heredoc. It's also not immediately clear, without aliasing the matcher `#expect_offense_and_correction`, that the second argument is an expected correction.

---

Then I was wondering if it's possible to leverage something like RSpec compound matchers. My understanding of these is severely lacking, however. I envisioned the API could be something like:

```ruby
expect_offense(<<-RUBY.strip_indent).and_no_correction
  foo
RUBY
```

This has the same limitation as the previous one though, with the need for another heredoc delimiter if an expected correction is passed as an argument.

---

I let my mind wander a bit, to try to find a solution to this, and thought up:

```ruby
inspect do
  expect_offense(<<-RUBY.strip_indent)
    foo
  end

  expect_correction(<<-RUBY.strip_indent)
    bar
  end

  # expect_no_correction
end
```

---

I'm not sure how feasible any of these ideas are, and there might be better alternatives that I just didn't think about. Hoping we can find a nice API for this as well. 🙂

**Comment by backus** [comment]
I've wanted this too. I've always thought it would be cool to have a diff like interface. For example

```ruby
expect_correction(<<~RUBY)
  def foo(bar)
@-    bar.map { |item| transform(item) }.flatten
@+    bar.flat_map { |item| transform(item) }
  end
RUBY
```

The problem is finding a nice way to annotate the diffs that isn't valid ruby syntax. Hence `@-` and `@+`

**Comment by backus** [comment]
BTW I really like your ideas here. The compound matchers and the `inspect do ... end` blocks are really neat solutions

**Comment by backus** [comment]
The more I think about it, the more I like the chained matcher like

```ruby
expect_offense(<<-RUBY).with_correction(<<-CORRECTION)
  def foo
    bar
    ^^^ Some offense
  end
RUBY
def foo
  baz
end
CORRECTION
```
only weirdness here is how ruby heredocs works. Not a huge deal since this is a ruby linter after all. People probably know Ruby decently well. We could also consider something like this though


expect_offense(<<-RUBY).with_correction
  def foo
    bar
    ^^^ Some offense
  end
  ~~~
  def foo
    baz
  end
RUBY
```

where it splits on the first line that is only tildes if that matcher is appended. Downside is it would have to play well with `strip_indent` unless the behavior was to split on `"~~~\n"` then do `strip_indent` but that isn't really generalizable. Also (and I feel kind of ridiculous saying this) I'm not sure if there is a case where those three tildes might be valid ruby.

**Comment by backus** [comment]
Also as side note, I find it sort of odd that most autocorrections are specified as the adjustment made by one cop even though, in practice, we are relying on how several cops all nudge the code in the right direction until the code is right. It makes me think about what this cop is doing in the first iteration and not what the code looks like to the end user. I'm not sure if involving other cops would be too "integration test-y" but it feels more natural to say "this is what I want the code to turn into if I wrote this code"

**Comment by backus** [comment]
Also, to be a bit more clear, I like this `.with_correction` solution better than my diff approach. I think the diff approach would be really confusing for asserting small whitespace adjustments.

**Comment by Drenmi** [comment]
> I'm not sure if involving other cops would be too "integration test-y" but it feels more natural to say "this is what I want the code to turn into if I wrote this code".

Yeah ... this touches on a more overarching problem, I guess. A lot of cops in the style department need a pass of auto-correct from the layout department after correcting. This is usually done implicitly, by re-running all other cops on the file after a correction.

But maybe there should be a declarative way of saying: "after this correction, run a pass of these layout cops."

This might feel like the cops know too much about each other, but on the other hand, quite a few cops already grab configuration from others to adjust their auto-correction, which is arguably much worse. We also already have `autocorrect_incompatible_with`, so we're sort of halfway down the line. 😅 

We've had a few two-three layers deep auto-correct chain bugs that result in syntax errors. Asserting the actual final source makes a lot of sense in that regard.

One problem I can think of is that other cops can have different configurations, too. So a decision needs to be made whether to only test the default configuration. Also, changes to the default configuration will require updating of test cases for unrelated cops. I don't necessarily see this as a deal breaker though. It's just a different trade-off. 🙂 

> I think the diff approach would be really confusing for asserting small whitespace adjustments.

Good point. I like it even though it is a bit esoteric. It is a good way of showing changes in git diffs. I think it's a clever way of looking at what others are already doing in analogous problem spaces. 🙂

**Comment by backus** [comment]
I don't know if some cops knowing the configuration of other cops is really the source of the smell we might be referring to here though. I think it might just be conceptually off to think of autocorrection as a per-cop thing.

It would be really hard to model and test (right now) what it would look like to say "these N cops perform autocorrect together" since that would probably result in an explosion of combinations. In that sense, specifying autocorrect per-cop makes sense because it seems to be the only thing that is conceptually simple enough to think about in unit tests.

---

I don't know exactly what the project organization would look like, but what if we had something like this:

```ruby
RuboCop::Spec.define do
  config do
    {
      'Layout/CopOne' => {
        'EnforcedStyle' => 'multiline'
      },
      'Layout/CopTwo' => {
        'EnforcedStyle' => 'pretty_style'
      }
    }
  end

  before <<-RUBY.strip_indent
    foo(bar, { baz:
    1, qux: 2})
  RUBY

  after <<-RUBY.strip_indent
    foo(
      bar,
      {
        baz: 1,
        qux: 2
      }
    )
  RUBY
end
```

in a separate file (sort of like how the meta directory works in mutant https://github.com/mbj/mutant/tree/master/meta). This would let us cleanly specify autocorrect behavior independent of a single cop.

More importantly though, we could have some special setup for these that make sure that we are able to go from **before** to **after** with some combination of cops. This setup could track the cops involved and expose that information. With the information of what cops work together with different autocorrects we could then add another layer of tests that then assert that autocorrect doesn't produce invalid source when you try each combination of config options for each cop involved.

Basically:

* it seems like specifying autocorrect independent of the cop that really just makes the first move would be helpful
* independent autocorrect specification would be more representative of what we are actually trying to do (tests as documentation and whatnot)
* independent autocorrect specification would let us possibly explore fuzzing different configuration combos to avoid complex autocorrect bugs

**Comment by bbatsov** [comment]
@rrosenblum is working on a solution for this in #6170, so hopefully we'll wrap this up soon.

**Comment by Darhazer** [comment]
#6170 is merged, introducing the `expect_correction` helper. I'm closing this issue.
